// ng-pwa-tools/bin/ngu-sw-manifest.js
#!/usr/bin/env node_modules/.bin/ts-node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var minimist = require("minimist");
var fs = require("fs");
var sw_manifest_1 = require("../lib/sw-manifest");
var util_1 = require("../lib/common/util");
var args = minimist(process.argv.slice(2), {
    string: ['in', 'module', 'base-href', 'dist', 'lazy-root', 'index', 'out'],
});
var dist = util_1.toAbsolute(args['dist'] || 'dist');
var appModule = args['module'] ? util_1.toAbsolute(args['module']) : undefined;
var loadChildrenRoot = util_1.toAbsolute(args['lazy-root'] || 'src');
var baseHref = args['base-href'] || '/';
var index = (baseHref.endsWith('/') ? baseHref : baseHref + '/') + (args['index'] || 'index.html');
var manifest = util_1.toAbsolute(args['in'] || 'src/ngsw-manifest.json');
manifest = fs.existsSync(manifest) ? manifest : undefined;
sw_manifest_1.generateSwManifest({
    appModule: appModule,
    baseHref: baseHref,
    dist: dist,
    index: index,
    loadChildrenRoot: loadChildrenRoot,
    manifest: manifest,
    routing: !!args['module'],
    static: true,
})
    .then(function (manifest) { return JSON.stringify(manifest, null, 2); })
    .then(function (manifest) {
    if (args['out']) {
        var out = util_1.toAbsolute(args['out']);
        fs.writeFileSync(out, manifest);
    }
    else {
        console.log(manifest);
    }
});
